File: app/src/main/java/com/baxolino/apps/floats/algorithms/AdlerFileWriter.java
package com.baxolino.apps.floats.algorithms;

import android.util.Log;

import androidx.annotation.NonNull;

import com.baxolino.apps.floats.core.Info;
import com.baxolino.apps.floats.core.io.BitStream;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Arrays;
import java.util.zip.Adler32;
import java.util.zip.GZIPOutputStream;

public class AdlerFileWriter {

  private static final String TAG = "AdlerFileWriter";

  private final InputStream input;

  private final OutputStream output;


  public AdlerFileWriter(InputStream input, OutputStream output) {
    this.input = input;
    this.output = output;
  }

  public void write() throws IOException {
    AdlerOutputStream adlerOutputStream = new AdlerOutputStream(output);
    GZIPOutputStream zipOutput = new GZIPOutputStream(adlerOutputStream);

    byte[] buffer = new byte[Info.BUFFER_SIZE];

    int read;
    while ((read = input.read(buffer)) > 0)
      zipOutput.write(buffer, 0, read);
    zipOutput.finish();

    long adlerCheckSum = adlerOutputStream.getValue();
    byte[] checksum = new BitStream()
            .writeLong64(adlerCheckSum)
            .toBytes();
    Log.d(TAG, "CheckSum = " + adlerCheckSum + " | " + Arrays.toString(checksum));

    output.write(
            new BitStream()
                    .writeLong64(adlerCheckSum)
                    .toBytes()
    );
    output.close();
    input.close();
  }
}

File: app/src/main/java/com/baxolino/apps/floats/algorithms/AdlerFileWriter.kt
package com.baxolino.apps.floats.algorithms

import android.util.Log
import com.baxolino.apps.floats.core.Info
import com.baxolino.apps.floats.core.io.BitStream
import java.io.IOException
import java.io.InputStream
import java.io.OutputStream
import java.util.Arrays
import java.util.zip.GZIPOutputStream

class AdlerFileWriter(private val input: InputStream, private val output: OutputStream) {

  fun write(listener: (Int) -> Unit) {
    val adlerOutputStream = AdlerOutputStream(output)

    val zipOutput = GZIPOutputStream(adlerOutputStream)
    val buffer = ByteArray(Info.BUFFER_SIZE)

    var read: Int
    var written = 0
    while (input.read(buffer).also { read = it } > 0) {
      zipOutput.write(buffer, 0, read)

      written += read
      listener.invoke(written)
    }
    zipOutput.finish()

    // send the checksum over to the other
    // device for verification; after sending bytes
    val adlerCheckSum = adlerOutputStream.value

    val checksum = BitStream()
      .writeLong64(adlerCheckSum)
      .toBytes()

    Log.d(TAG, "CheckSum = " + adlerCheckSum + " | " + Arrays.toString(checksum))

    output.write(
      BitStream()
        .writeLong64(adlerCheckSum)
        .toBytes()
    )
    output.close()
    input.close()
  }

  companion object {
    private const val TAG = "AdlerFileWriter"
  }
}